File: 1475.py
class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        res = list()
        firstSmallIdxAfter = [0] * len(prices)
        res.append(prices[-1])
        firstSmallIdxAfter[-1] = len(prices)
        for i in range(len(prices) - 2, -1, -1):
            ptr = i + 1
            while ptr < len(prices) and prices[i] < prices[ptr]:
                ptr = firstSmallIdxAfter[ptr]
            firstSmallIdxAfter[i] = ptr
            if ptr != len(prices):
                res.append(prices[i] - prices[firstSmallIdxAfter[i]])
            else:
                res.append(prices[i])
        return res[::-1]

File: sol/1475.cpp
struct Solution {
    vector<int> finalPrices(const vector<int>& prices) {
        size_t n = prices.size(), ptr;
        short i;
        vector<int> res, smallIdxAfter;
        res.resize(n);
        smallIdxAfter.resize(n);
        res[n - 1] = prices[n - 1];
        smallIdxAfter[n - 1] = n;
        for (i = n - 2; i >= 0; --i){
            ptr = i + 1;
            while (ptr < n && prices[i] < prices[ptr]){
                ptr = smallIdxAfter[ptr];
            }
            smallIdxAfter[i] = ptr;
            if (ptr == n){
                res[i] = prices[i];
            } else {
                res[i] = prices[i] - prices[smallIdxAfter[i]];
            }
        }
        return res;
    }
};
